db/external_sst_file_test.cc

```diff
@@ -2063,6 +2080,89 @@ TEST_F(ExternalSSTFileTest, LinkExternalSst) {
   }
 }
 
+/*
+ * Test and verify the functionality of ingestion_options.move_files and ingestion_options.failed_move_fall_back_to_copy
+ */
+TEST_F(ExternalSSTFileTest, LinkFailFallBackExternalSst) {
+  Options options = CurrentOptions();
+  options.disable_auto_compactions = true;
+  DestroyAndReopen(options);
+  const int kNumKeys = 10000;
+
+  std::string file_path = sst_files_dir_ + "file1.sst";
+  // Create SstFileWriter for default column family
+  SstFileWriter sst_file_writer(EnvOptions(), options);
+  ASSERT_OK(sst_file_writer.Open(file_path));
+  for (int i = 0; i < kNumKeys; i++) {
+    ASSERT_OK(sst_file_writer.Put(Key(i), Key(i) + "_value"));
+  }
+  ASSERT_OK(sst_file_writer.Finish());
+  uint64_t file_size = 0;
+  ASSERT_OK(env_->GetFileSize(file_path, &file_size));
+
+  // Failed move falls back to copy.
+    IngestExternalFileOptions ifo;
+    ifo.move_files = true;
+    ifo.failed_move_fall_back_to_copy = true;
+    ExternalSSTTestEnv test_env(env_, true);
+    InstrumentedMutex mutex_;
+    std::list<uint64_t>::iterator pending_output_elem;
+    const uint64_t next_file_number =
+        dbfull()->Test_GetVersionSet()->FetchAddFileNumber(1);
+
+    ColumnFamilyData* cfd =
+        static_cast<ColumnFamilyHandleImpl*>(db_->DefaultColumnFamily())->cfd();
+    ExternalSstFileIngestionJob job(&test_env, dbfull()->Test_GetVersionSet(),
+                                    cfd, dbfull()->immutable_db_options(),
+                                    dbfull()->Test_GetEnvOptions(), nullptr,
+                                    ifo);
+    SuperVersion* super_version = cfd->GetReferencedSuperVersion(&mutex_);
+    ASSERT_OK(job.Prepare({file_path}, next_file_number + 1, super_version));
+    dbfull()->CleanupSuperVersion(super_version);
+    // Copy file is true since a failed link falls back to copy file.
+    ASSERT_TRUE(job.files_to_ingest()[0].copy_file);
+}
+
+TEST_F(ExternalSSTFileTest, LinkFailNoFallBackExternalSst) {
```

Comment:
I suggest that we use parameterized test for both cases to reduce code duplication (we may need a separate class because ExternalSstFileTest is already parameterized). Furthermore, you can use `TEST_SYNC_POINT_CALLBACK` to expose certain information from internal function implementations to the unit tests. An example would be https://github.com/facebook/rocksdb/blob/master/table/block_based_table_reader.cc#L899.

Reply:
Thanks. I changed the test to use TEST_P and added TEST_SYNC_POINT_CALLBACK.